### PR TITLE
Sugarchain: remove: disable additional download traffic (more getheaders) during IBD

### PR DESCRIPTION
**Pull request by decryp2kanon**
https://github.com/sugarchain-project/sugarchain/issues/45

### before
image

### after
image

### comparision
blue vs purple
image

### CODE REVIEW

**Comment by decryp2kanon** [comment]
`peer eviction timer should be less than stale tip check timer`
```bash
net_processing.cpp: In constructor ‘PeerLogicValidation::PeerLogicValidation(CConnman*, CScheduler&)’:
net_processing.cpp:808:5: error: static assertion failed: peer eviction timer should be less than stale tip check timer
     static_assert(EXTRA_PEER_CHECK_INTERVAL < STALE_CHECK_INTERVAL, "peer eviction timer should be less than stale tip check timer");
     ^
Makefile:5792: recipe for target 'libbitcoin_server_a-net_processing.o' failed
```

**Comment by decryp2kanon** [comment]
```
====================================================
   Sugarchain Yumekawa 0.16.3: src/test-suite.log
====================================================

# TOTAL: 2
# PASS:  1
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: test/test_sugarchain
==========================

Running 282 test cases...
test/DoS_tests.cpp(138): error in "stale_tip_peer_management": check connman->GetTryNewOutboundPeer() failed
test/DoS_tests.cpp(155): error in "stale_tip_peer_management": check vNodes.back()->fDisconnect == true failed
test/DoS_tests.cpp(167): error in "stale_tip_peer_management": check vNodes[nMaxOutbound-1]->fDisconnect == true failed

*** 3 failures detected in test suite "Sugarchain Test Suite"
FAIL test/test_sugarchain (exit status: 201)
```

**Comment by decryp2kanon** [comment]
image

**Comment by decryp2kanon** [comment]
image

still no effect

**Comment by decryp2kanon** [comment]
image

**Comment by decryp2kanon** [comment]
image

**Comment by decryp2kanon** [comment]
`socket recv error Connection reset by peer (104)`

**Comment by decryp2kanon** [comment]
image

60% better

**Comment by decryp2kanon** [comment]
image
no effect

**Comment by decryp2kanon** [comment]
image
no effect

**Comment by decryp2kanon** [comment]
actual size = 17 MB
displayed size = 2048 MB (`120x` greater)

**Comment by cryptozeny** [comment]
` -blocksonly`

**Comment by decryp2kanon** [comment]
`net_processing.cpp`
```
MarkBlockAsInFlight
MarkBlockAsReceived
```

**Comment by decryp2kanon** [comment]
> warning: integer overflow in expression [-Woverflow]

**Comment by decryp2kanon** [comment]
https://github.com/stratisproject/StratisBitcoinFullNode/pull/1158

**Comment by decryp2kanon** [comment]
bool IsInitialBlockDownload()

**Comment by decryp2kanon** [comment]
little (1 hr) faster, but no clue
image

**Comment by decryp2kanon** [comment]
`socket send error Broken pipe (32)`

**Comment by decryp2kanon** [comment]
image
no effect

**Comment by decryp2kanon** [comment]
`nBlocksInFlight`

**Comment by decryp2kanon** [comment]
```cpp
                // Download as much as possible, from earliest to latest.
                for (const CBlockIndex *pindex : reverse_iterate(vToFetch)) {
                    if (nodestate->nBlocksInFlight * 120 >= MAX_BLOCKS_IN_TRANSIT_PER_PEER) { // FIXME.SUGAR // 120x bitcoin
                        // Can't download any more from this peer
                        break;
                    }
                    uint32_t nFetchFlags = GetFetchFlags(pfrom);
                    vGetData.push_back(CInv(MSG_BLOCK | nFetchFlags, pindex->GetBlockHash()));
                    MarkBlockAsInFlight(pfrom->GetId(), pindex->GetBlockHash(), pindex);
                    LogPrint(BCLog::NET, "Requesting block %s from  peer=%d\n",
                            pindex->GetBlockHash().ToString(), pfrom->GetId());
                }
```
net_processing.cpp

**Comment by decryp2kanon** [comment]
2020-02-12 11:43:49 getheaders (6000) 96f3fda6c2da607766d3592a0a76f525d34d6935d7f99efbb2c6e40e734120b9 to peer=3
2020-02-12 11:43:49 getheaders (6000) 96f3fda6c2da607766d3592a0a76f525d34d6935d7f99efbb2c6e40e734120b9 to peer=2
2020-02-12 11:43:49 getheaders (6000) 96f3fda6c2da607766d3592a0a76f525d34d6935d7f99efbb2c6e40e734120b9 to peer=1
2020-02-12 11:43:49 getheaders (6000) 96f3fda6c2da607766d3592a0a76f525d34d6935d7f99efbb2c6e40e734120b9 to peer=0
```

**Comment by decryp2kanon** [comment]
```
tail -f -n +1 ~/.sugarchain3-test/debug.log | grep -Ev "more|sending|received:" | grep "getheaders ("
```

**Comment by decryp2kanon** [comment]
it seems duplicated
image

**Comment by decryp2kanon** [comment]
```cpp
        if (nCount == MAX_HEADERS_RESULTS) {
            // Headers message had its maximum size; the peer may have more headers.
            // TODO: optimize: if pindexLast is an ancestor of chainActive.Tip or pindexBestHeader, continue
            // from there instead.
            LogPrint(BCLog::NET, "more getheaders (%d) to end to peer=%d (startheight:%d)\n", pindexLast->nHeight, pfrom->GetId(), pfrom->nStartingHeight);
            connman->PushMessage(pfrom, msgMaker.Make(NetMsgType::GETHEADERS, chainActive.GetLocator(pindexLast), uint256()));
        }
```

**Comment by decryp2kanon** [comment]
image
`21M`

**Comment by decryp2kanon** [comment]
still duplicated, but its much better.
image

**Comment by decryp2kanon** [comment]
its fine.
image

**Comment by decryp2kanon** [comment]
purple
image

**Review by cryptozeny — APPROVED** [review]
Tested ACK ebc5c9a3ce8503609332c48c49cbae5fb24f574b

**Comment by decryp2kanon** [comment]
https://github.com/bitcoin/bitcoin/pull/8306

**Comment by decryp2kanon** [comment]
https://github.com/bitcoin/bitcoin/pull/9091/files

File: src/net_processing.cpp [hunk 1]
```diff
@@ -3125,7 +3129,9 @@ void PeerLogicValidation::CheckForStaleTipAndEvictPeers(const Consensus::Params
         // Check whether our tip is stale, and if so, allow using an extra
         // outbound peer
         if (TipMayBeStale(consensusParams)) {
-            LogPrintf("Potential stale tip detected, will try using extra outbound peer (last tip update: %d seconds ago)\n", time_in_seconds - g_last_tip_update);
+            if (!IsInitialBlockDownload()) { // FIXME.SUGAR // LOG_DISABLED during IBD
+                LogPrintf("Potential stale tip detected, will try using extra outbound peer (last tip update: %d seconds ago)\n", time_in_seconds - g_last_tip_update);
+            }
             connman->SetTryNewOutboundPeer(true);
```
Comment (decryp2kanon) [inline src/net_processing.cpp]: test if this disabled during IBD

**Pull request merged**